[Merged by Bors] - Consensus context with proposer index caching

michaelsproul opened the pull request:
## Issue Addressed

Closes https://github.com/sigp/lighthouse/issues/2371

## Proposed Changes

Backport some changes from `tree-states` that remove duplicated calculations of the `proposer_index`.

With this change the proposer index should be calculated only once for each block, and then plumbed through to every place it is required.

## Additional Info

In future I hope to add more data to the consensus context that is cached on a per-epoch basis, like the effective balances of validators and the base rewards.

There are some other changes to remove indexing in tests that were also useful for `tree-states` (the `tree-states` types don't implement `Index`).

michaelsproul committed:
ConsensusContext ported to unstable

michaelsproul committed:
More proposer index plumbing

michaelsproul commented:
This is complementary to https://github.com/sigp/lighthouse/pull/3263, but would likely conflict slightly syntactically.

I'll also post some benchmark numbers soon (TM)

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3645,12 +3645,13 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             ProduceBlockVerification::VerifyRandao => BlockSignatureStrategy::VerifyRandao,
             ProduceBlockVerification::NoVerification => BlockSignatureStrategy::NoVerification,
         };
+        let mut ctxt = ConsensusContext::new(block.slot()).set_proposer_index(proposer_index);
```
On second thoughts we probably shouldn't set the proposer index during block production, as verifying it is desired (see: https://github.com/sigp/lighthouse/pull/2740)

The context will take care of calculating it only once

michaelsproul committed:
Block production checks, tweak lcli

michaelsproul commented:
Benchmarks indicate a 7.5-20% reduction in run time for block processing for a recent block (slot 4769389).

The reduction was greatest on my Apple M1 Pro, from a median time of 38.2ms to 30.9ms. On an AMD Ryzen 5950X the reduction was from 40.8ms to 34.7ms.

Full details in this spreadsheet: https://docs.google.com/spreadsheets/d/1axT3Tda3wzWXdUzPEt5rAmxWE7PdIkSta7sW5PHfdq0/edit

michaelsproul committed:
Fix some beacon chain tests

paulhauner reviewed (approved):
Very nice, I like the `ConsensusContext`! I notice you didn't go with `ctx` :wink: I have a couple of comments, but they're not strictly necessary!

I had a lil' grep and noticed two places where we're still computing the proposer index, perhaps you're aware of them:

- `consensus/state_processing/src/common/slash_validator.rs:    let proposer_index = state.get_beacon_proposer_index(state.slot(), spec)?;`
- `consensus/state_processing/src/per_block_processing/process_operations.rs:        let proposer_index = state.get_beacon_proposer_index(state.slot(), spec)? as u64;`
    - This is phase0, so there are marginal returns from this. Perhaps useful during sync, for what that's worth in a WSS world. I could take or leave this.
    
Additionally, I believe this call is totally redundant and could be removed (it could also happen in another PR, potential scope creep):

https://github.com/sigp/lighthouse/blob/01e84b71f524968f5b940fbd2fa31d29408b6581/beacon_node/beacon_chain/src/beacon_chain.rs#L3501

michaelsproul committed:
Remove redundant proposer index calc in block production

michaelsproul committed:
Remove remaining proposer index recalculations

michaelsproul commented:
Thanks for the review Paul!

I've addressed all your comments in these two commits:

- Removing the re-calc in block production: e1e1e89a62fc03c59bf733d2cfa9a2c09be19bb4
- Removing the re-calc in block processing: bb7e88e160035d6059ce9aaa68633c1789e08bec

Figured we may as well try to be fast for block replay and in case of mass slashings! Nice catch

michaelsproul committed:
Omit proposer recalculation from block replay

michaelsproul commented:
Pushed one (1) more sneaky commit with an optimisation for block replay: 26b819f7f88b8dad5cf3d5b19cdb6438a22f3ba4. I was going to put it in a follow-up but figured I'd streamline things by including it here.

paulhauner reviewed (approved):
Perfect, sorry for the review delay!

michaelsproul commented:
Thank you for the review!

bors r+
